src/walletConnect/request.ts

```diff
@@ -4,33 +4,134 @@ import { ContractKit } from '@celo/contractkit'
 import { UnlockableWallet } from '@celo/wallet-base'
 import { SentryTransactionHub } from 'src/sentry/SentryTransactionHub'
 import { SentryTransaction } from 'src/sentry/SentryTransactions'
+import { getFeatureGate } from 'src/statsig'
+import { StatsigFeatureGates } from 'src/statsig/types'
 import { chooseTxFeeDetails, sendTransaction } from 'src/transactions/send'
 import { newTransactionContext } from 'src/transactions/types'
+import { estimateFeesPerGas } from 'src/viem/estimateFeesPerGas'
+import { ViemWallet } from 'src/viem/getLockableWallet'
 import { SupportedActions } from 'src/walletConnect/constants'
-import { getContractKit, getWallet, getWeb3 } from 'src/web3/contracts'
+import { getContractKit, getViemWallet, getWallet, getWeb3 } from 'src/web3/contracts'
+import networkConfig from 'src/web3/networkConfig'
 import { getWalletAddress, unlockAccount } from 'src/web3/saga'
 import { applyChainIdWorkaround, buildTxo } from 'src/web3/utils'
 import { call } from 'typed-redux-saga'
+import {
+  Address,
+  GetTransactionCountParameters,
+  SignMessageParameters,
+  formatTransaction,
+} from 'viem'
+import { getTransactionCount } from 'viem/actions'
 import Web3 from 'web3'
 
 const TAG = 'WalletConnect/handle-request'
 
-export interface WalletResponseError {
-  isError: true
-  error: string
-}
-export interface WalletResponseSuccess {
-  isError: false
-  result: string
-}
-
 export function* handleRequest({ method, params }: { method: string; params: any[] }) {
-  const account: string = yield* call(getWalletAddress)
-  const wallet: UnlockableWallet = yield* call(getWallet)
+  const wallet: ViemWallet = yield* call(getViemWallet, networkConfig.viemChain.celo)
+  const useViem = yield* call(
+    getFeatureGate,
+    StatsigFeatureGates.USE_VIEM_FOR_WALLETCONNECT_TRANSACTIONS
+  )
+
+  const account = yield* call(getWalletAddress)
+  const legacyWallet: UnlockableWallet = yield* call(getWallet)
   yield* call(unlockAccount, account)
   // Call Sentry performance monitoring after entering pin if required
   SentryTransactionHub.startTransaction(SentryTransaction.wallet_connect_transaction)
 
+  // TODO: keep only Viem branch after feeCurrency estimation is ready
+  if (useViem) {
+    switch (method) {
+      case SupportedActions.eth_signTransaction: {
+        const rawTx: any = { ...params[0] }
+
+        let tx: any
+        // Provide an escape hatch for dapp developers who don't want any normalization
+        if (rawTx.__skip_normalization) {
+          delete rawTx.__skip_normalization
+          tx = rawTx
+        } else {
+          tx = yield* call(normalizeTransaction, wallet, rawTx)
+        }
+
+        // Convert hex values to numeric ones for Viem
+        // TODO: remove once Viem allows hex values as quanitites
+        const formattedTx: any = yield* call(formatTransaction, tx)
```

Comment:
can the bulk of this logic be shared between sendTransaaction and signTransaction? (the only difference i can see, aside from the obvious wallet method that's called at the very end, is that we don't respect `__skip_normalization` for sendTransaction) is it intentional that the logic will diverge between these 2 methods at some point so we haven't abstracted the logic for formatting the tx?

Reply:
probably we don't need `__skip_normalization` at all, wdyt @jeanregisser? why someone could want to force push forward a wrong transaction?

Reply:
Yes sounds good to remove it 👍

Reply:
great, I've abstracted all shared logic about preparing tx request into a separate function